lambdas/shared/t4_lambda_shared/preview.py

```diff
@@ -129,81 +129,80 @@ def extract_fcs(file_, as_html=True):
+
     Returns:
-        dict
+        tuple
             body - summary of main contents (if applicable)
-            info - metdata for user consumption
+            info - metadata for user consumption
     """
-    # TODO: generalize to datasets, multipart files
-    # As written, only works for single files, and metadata
-    # is slanted towards the first row_group
-    # local import reduces amortized latency, saves memory
+    logger_ = get_quilt_logger()
     import pyarrow.parquet as pq  # pylint: disable=C0415
 
     pf = pq.ParquetFile(file_)
     meta = pf.metadata
 
+    body = ""
     info = {}
     info['created_by'] = meta.created_by
     info['format_version'] = meta.format_version
     info['note'] = TRUNCATED
     info['num_row_groups'] = meta.num_row_groups
     # in previous versions (git blame) we sent a lot more schema information
     # but it's heavy on the browser and low information; just send column names
-    info['schema'] = {
-        'names': meta.schema.names
-    }
-    info['serialized_size'] = meta.serialized_size
+    info['schema'] = {'names': meta.schema.names}
+    info['serialized_size'] = meta.serialized_size  # footer serialized size
     info['shape'] = [meta.num_rows, meta.num_columns]
-    # TODO: refactor preview code to use dask/s3fs and pyarrow.dataset scanner to
-    # spare memory; part of the reason this is so inefficient: we've already read
-    # the entire parquet file into a BytesIO by the time we get here
-    if meta.num_row_groups:
-        # guess because we meta doesn't reveal how many rows in first group
-        num_rows_guess = math.ceil(meta.num_rows / meta.num_row_groups)
-        size_guess = num_rows_guess * meta.num_columns * AVG_PARQUET_CELL_BYTES
-        if skip_rows or (size_guess > get_available_memory()):
-            # minimal dataframe with all columns and one row
-            dataframe = pandas.DataFrame(columns=meta.schema.names)
-            info['warnings'] = 'Large file: skipped rows to conserve memory, only showing column names'
-        else:
-            dataframe = pf.read_row_group(0)[:MAX_PREVIEW_ROWS].to_pandas()
-    # sometimes there are neither rows nor row_groups, just columns
-    # therefore we do not call read_row_group because (with 0 row_groups)
-    # it would barf
-    else:
-        # :0 is a safety valve but there really should be no rows in this case
-        dataframe = pf.read()[:0].to_pandas()
-    if as_html:
-        body = remove_pandas_footer(dataframe._repr_html_())  # pylint: disable=protected-access
+
+    available = get_available_memory()
+    iter_batches = None
+    # 10MB heuristic; should never happen, e.g. with current default of 512MB
+    if (available < 10E6) or skip_rows:
+        logger_.warning("Insufficient memory to index parquet file: %s", info)
+        info['warnings'] = "Skipped rows; insufficient memory"
+    elif meta.num_rows and meta.num_row_groups:
```

Comment:
Is `num_row_groups` still relevant? You don't call `read_row_group` anymore.

I'm trying to understand if batches and row groups are related - but I don't think they are? So you could always use `iter_batches`?

Reply:
It is because it has to be non-zero or this `row_groups` argument will fail: `iter_batches = pf.iter_batches(batch_size=128, row_groups=[0])`.

Reply:
In theory num_rows > 0 implies num_row_groups > 0 but I don't see a good reason to take a chance. Of course num_row_groups > 0 does not imply num_rows > 0.